packages/apollo-language-server/src/utilities/graphql.ts

```diff
@@ -103,6 +113,18 @@ export function removeDirectiveAnnotatedFields(
         )
       )
         return null;
+
+      // We're keeping this field. Save the names of all fragment spreads that are in this field's selection
+      // set. This is intentionally _not_ recursive because descendents of this field may a directive that we
+      // want to remove; so we only know that this field's selection set is safe to keep.
+      if (node.selectionSet) {
+        node.selectionSet.selections.forEach(selection => {
```

Comment:
Instead of looking through selection sets (which would end up happening recursively due to the visitor btw as all `Fields` would be visited), we should visit on locations where fragments can be present to determine if they are impacted by the `@client` directive. The three locations can be found [here](https://github.com/graphql/graphql-js/blob/35c68d403cf1020e1d30ea29f0546352d03dd49e/src/language/kinds.js#L19-L21).

The other part of the story (which you are doing below!) is to trim the fragments that are no longer used) after collecting the impacted ones. Instead of checking for selection > @client > fragment spread, we should do a general check at the end for fragments orphaned by the trimming of selection sets done through field splitting.

Note that this will require doing a recursive lookup for fragments that could be impacted by the client directive (prior art https://github.com/graphql/graphql-js/blob/5c9946b0d0e2e082772db4ccd50013caa0b1ed79/src/validation/ValidationContext.js#L110)

```graphql
fragment One on Node {
  ...Two
  user {
    friends {
       name
      ...Two @client 
    }
  }
}
fragment Two on Node {
  id
}

query {
  me {
    ...One
  }
}
```

In the above example we *don't* want to trim `Two` as it is still used even though we *do* want to trim the usage of `Two` on `friends`.

In a similar issue, if a fragment is annotated as `@client` and contains fragments, we need to recurse and trim those fragments *if* they aren't used elsewhere  in the operation (i.e.):

```graphql
fragment One on Node @client {
  ... Two
}
fragment Two on Node {
  id
}
query {
  me { name ...One }
}
```

In the above query, all that should be left in the document is `query { me { name } }` as both `One` and `Two` should be trimmed by the removal